### PR TITLE
Accept header negotiation for the output format.

### PR DESCRIPTION
**Pull request by e0ipso**
If several output formats are available, allow clients to select one using the `Content-Type` header.

### CODE REVIEW

**Comment by e0ipso** [comment]
Marking as "release blocker" since I don't expect this to be too hard and will provide a cool feature.

**Comment by e0ipso** [comment]
@amitaibu tests are back green including new tests.

**Comment by amitaibu** [comment]
entity-view tests seem to be commented out

**Comment by e0ipso** [comment]
Fuck, you're right.

File: tests/RestfulViewEntityTestCase.test [hunk 1]
```diff
@@ -31,216 +31,248 @@ class RestfulViewEntityTestCase extends RestfulCurlBaseTestCase {
...
+    ));
+    $result = drupal_json_decode($response['body']);
+    $this->assertNotNull($result, 'JSON output detected.');
+
+    // Test XML selection.
+    $response = $this->httpRequest('api/v1/articles/' . $node->nid, \RestfulInterface::GET, NULL, array(
+      'Accept' => 'application/xml',
+    ));
+    $result = new SimpleXMLElement($response['body']);
+    $this->assertNotNull($result, 'JSON output detected.');
```
Comment (amitaibu) [inline tests/RestfulViewEntityTestCase.test]: => XML

**Comment by e0ipso** [comment]
ready

File: restful.module [hunk 2]
```diff
@@ -722,6 +750,71 @@ function restful_delivery($var = NULL, $method = 'format') {
...
+ * Helper function to get the default output format from the current request.
+ *
+ * @param \RestfulBase $restful_handler
+ *   The restful handler for the formatter.
+ *
+ * @return \RestfulFormatterBase
+ *   The formatter plugin to use.
+ */
+function restful_output_format(\RestfulBase $restful_handler = NULL) {
+  if ($restful_handler || $restful_handler = restful_get_restful_handler_for_path()) {
```
Comment (amitaibu) [inline restful.module]: Maybe a bit cleaner:

``` php
$restful_handler = $restful_handler ? $restful_handler : restful_get_restful_handler_for_path();
if ($restful_handler && $formatter_name = $restful_handler->getPluginInfo('formatter')) {
  return restful_get_formatter_handler($formatter_name, $restful_handler);
}
```

File: restful.module [hunk 3]
```diff
@@ -722,6 +750,71 @@ function restful_delivery($var = NULL, $method = 'format') {
...
+ *
+ * @see drupal_match_path().
+ */
+function restfult_match_content_type($content_type, $pattern) {
+  $regexps = &drupal_static(__FUNCTION__);
+
+  if (!isset($regexps[$pattern])) {
+    // Convert path settings to a regular expression.
+    // Therefore replace newlines with a logical or, /* with asterisks and the <front> with the frontpage.
+    $to_replace = array(
```
Comment (amitaibu) [inline restful.module:802]: I'm a bit confused by the regex - maybe you can add in the comment an example(s)

File: tests/RestfulViewEntityTestCase.test [hunk 4]
```diff
@@ -235,12 +235,42 @@ class RestfulViewEntityTestCase extends RestfulCurlBaseTestCase {
   /**
    * Test the generation of the Vary headers.
    */
-  public function testVaryHeader() {
+  public function testHeaders() {
+    $node = $this->drupalCreateNode(array('type' => 'article', 'status' => NODE_PUBLISHED));
```
Comment (amitaibu) [inline tests/RestfulViewEntityTestCase.test]: I think `'status' => NODE_PUBLISHED` is already by default.

File: tests/RestfulViewEntityTestCase.test [hunk 5]
```diff
@@ -235,12 +235,42 @@ class RestfulViewEntityTestCase extends RestfulCurlBaseTestCase {
...
     // When there is no minor version passed in there is no need of Vary.
     $response = $this->httpRequest('api/v1/articles');
     $this->assertFalse(preg_match('/X-Restful-Minor-Version/', $response['headers']), 'The Vary header was not added.');
 
     $response = $this->httpRequest('api/v1/articles', \RestfulInterface::GET, NULL, array('X-Restful-Minor-Version' => 1));
     $this->assertTrue(preg_match('/X-Restful-Minor-Version/', $response['headers']), 'The Vary header was added.');
+
+    // Test that if there is no explicit formatter in the plugin definition then
+    // it is selected based on the Accept header.
+    variable_set('restful_default_output_formatter', 'invalid');
```
Comment (amitaibu) [inline tests/RestfulViewEntityTestCase.test:250]: maybe also worth testing accept headers on `articles__1_5` that explicetly defines the XML formatter.

File: restful.module (hunk 2 above)
Comment (e0ipso) [inline restful.module]: neat!

**Comment by e0ipso** [comment]
@amitaibu done.

**Comment by amitaibu** [comment]
@mateu-aguilo-bosch can you please check the 2 other comments.

**Comment by e0ipso** [comment]
Not sure what you mean. Those were taken care of as well.

File: restful.module (hunk 3 above)
Comment (amitaibu) [inline restful.module:802]: > `/* with asterisks and the <front> with the frontpage.`

^^ I'm not clear about this comment.

File: tests/RestfulViewEntityTestCase.test (hunk 5 above)
Comment (amitaibu) [inline tests/RestfulViewEntityTestCase.test:250]: @mateu-aguilo-bosch ^^ - do you agree on this?

Comment (e0ipso) [inline tests/RestfulViewEntityTestCase.test:250]: Sure: https://github.com/Gizra/restful/pull/155/files#diff-03d7caba797b37ec45fa863e6fad3033R270

**Comment by amitaibu** [comment]
Merged, thanks!

**Pull request closed without merging**